### ISSUE TITLE
RFC: Confidence calibration — formalize finding confidence semantics

### ISSUE THREAD

**Issue by autholykos** [dusk-network/pituitary#171]
```

ccd-backlog
id = 171
kind = "item"
section = "architecture"
status = "ready"
effort = "large"
impact = "high"
priority_rank = 250
depends_on = []
```

## Context

Pituitary reports confidence levels (high/medium/low) with numeric scores for drift findings, overlap detection, and compliance checks. Currently these thresholds are ad hoc — set during development and not formally calibrated.

## Questions this RFC should answer

- What do high/medium/low confidence mean in concrete terms?
- Should teams block merges on medium-confidence drift findings?
- How should thresholds differ between fixture (deterministic) and semantic (LLM) modes?
- Should users be able to configure confidence thresholds?
- How do we avoid false confidence — the "green means safe" trap?

## Why now

As teams rely on Pituitary in CI (Segment 2), they need to trust the numbers. RFC 0001 guardrail: "The product must abstain when coverage is insufficient." This RFC defines what "insufficient" means numerically.

**Comment by autholykos** [comment]
## Empirical data point — Qwen3.5-9B analysis pass, 2026-04-10

Full consistency run on a ~26-spec documentation corpus (58 artifacts, 643 chunks, 300 edges with AST inference enabled per #288) using Qwen3.5-9B as the `runtime.analysis` model, served locally via an OpenAI-compatible MLX endpoint. `pituitary 1.0.0-beta.8`, embeddings via `nomic-embed-text-v1.5`.

### Observation: Qwen3.5-9B produced high-confidence false positives at 0.95

Two findings in this run were reported with `confidence_score ≈ 0.95` that ground-truth verification against the actual source files showed to be incorrect.

**False positive 1 — check-doc-drift**
- Finding: a CHANGELOG block was flagged as drifting from a spec governing runtime state.
- Reality: the CHANGELOG section is *past-tense narration* of the release that implemented the spec-mandated change. The spec body explicitly states that the legacy derived file referenced in the CHANGELOG is not part of the accepted runtime contract. The CHANGELOG and the spec agree — the CHANGELOG just describes the historical moment when they started agreeing.
- The model keyed on surface-level term co-occurrence (the filename appearing in both texts with different framing) and missed the pragmatic register.

**False positive 2 — check-compliance**
- Finding: a reference doc was flagged with a `semantic_conflict` between two distinct CLI subcommands that happen to share the leading verb (think `foo start --activate` vs `foo session-state start --mode <mode>`).
- Reality: both are real, independent subcommands with their own flag surfaces. The reference doc correctly cites the right one for its context. There is no conflict.
- The model treated two disjoint CLI subcommands as if they were the same surface.

### What this suggests for the RFC

1. **Confidence semantics should distinguish statistical vs. pragmatic judgments.** On the same run, the terminology check was near-flawless (44 findings / 366 matches, all verified against the `[[terminology.policies]]` block — this is a pattern-match task with policy resolution, no reasoning about intent needed). Drift and conflict detection, by contrast, require reading *why* a sentence is written the way it is, and that's where smaller models (9B class) hit a ceiling.
2. **Consider a model-class factor in the confidence score.** A 0.95 from a 9B model on a drift adjudication probably shouldn't be directly comparable to a 0.95 from a 35B model on the same task, even if the raw logits agree. One framing: `effective_confidence = raw_confidence × task_class_factor(model_size, task_type)` where `task_class_factor` for `(9B, drift)` is penalized vs. `(35B, drift)`.
3. **Two-stage adjudication as a product pattern.** The heuristic that shook out for us on this corpus: *trust the 9B on statistical findings, re-adjudicate pragmatic judgments against a 35B before acting on any `confirmed_*` or `semantic_conflict` severity.* That's an ergonomic workflow that the confidence score, on its own, doesn't currently support — today there's nothing that tells you "this kind of finding at this confidence from this model is risky enough to re-run."
4. **Publish a calibration corpus.** Something like a small, versioned, labeled set of known-good and known-bad findings that users can run against their chosen analysis model before trusting its confidence scores on their own corpus. The RFC could define the format and acceptance criteria.

Happy to share the full JSON output of both false-positive findings if useful — they're compact and make good regression fodder.

### Setup notes

The local MLX router exposes both Qwen3.5-9B and Qwen3.5-35B endpoints behind an OpenAI-compatible API; the results above are specifically about the 9B. A 35B re-run on the same corpus is the next thing I'm doing and I'll post that comparison when it's done.

**Comment by autholykos** [comment]
## Follow-up: Qwen3.5-35B re-adjudication on the same corpus

Re-ran the two analysis checks that previously produced false positives — this time with the Qwen3.5-35B endpoint in `[runtime.analysis]`, everything else identical (same index, same chunks, same `nomic-embed-text-v1.5` embeddings).

### Side-by-side

| | Qwen3.5-9B | Qwen3.5-35B |
|---|---|---|
| `check-compliance` compliant | 1 | 1 |
| `check-compliance` **conflicts** | **1 false positive @ 0.95** | **0** ✅ |
| `check-compliance` unspecified | 25 | 25 |
| `check-doc-drift` confirmed drift | 1 @ 0.95 | 1 @ 0.834 |
| `check-doc-drift` possible_drift | 29 | 29 (all labeled level=low) |
| `check-doc-drift` aligned | 1 | 1 |

### What the 35B got right that the 9B got wrong

The `semantic_conflict` on the reference doc — the 9B flagged it at confidence 0.95, claiming that the two CLI subcommands contradict each other. **The 35B correctly did not flag this** — they're distinct subcommands that do different things. The larger model preserved the pragmatic distinction between two command surfaces that share some tokens.

### What the 35B got wrong

The 35B still produced 1 false positive on `check-doc-drift` — but on a **different** CHANGELOG section than the 9B. Both models failed on the same class of problem: historical narration in CHANGELOG entries.

**The 35B's finding** (confidence 0.834, level=high):
- Section: a `Changelog / [0.6.2] - 2026-03-18 / Changed` entry
- Cited excerpt: *"**Native-canonical runtime state.** Clone-local runtime truth now lives in `runtime_state/state.json`, while `foo.md` is treated as a derived human-readable export instead of a peer source of runtime truth."*
- Rationale: *"accepted spec says foo.md is not part of the active runtime contract, but the doc still presents it as active runtime state."*

The CHANGELOG entry is literally the demotion ("**instead of** a peer source of runtime truth"), consistent with the spec. The model pattern-matched on token proximity and missed the "instead of" structural clue — same class of failure as the 9B, just at lower confidence and on a different section.

### What this suggests for the RFC

1. **Model class correlates with some failure types but not others.** Pragmatic disambiguation between CLI subcommands (or between two things that share vocabulary but do different jobs) benefits clearly from scaling up — the 35B got 100% of the 9B's compliance false positives right. But historical-container narration failures (CHANGELOG, release notes, migration docs) are shared across model sizes. A bigger model isn't the fix there; corpus-aware prompting or a deterministic post-filter is.

2. **The 35B's confidence layer is actually well-calibrated already.** It reports `confidence: {level, score, basis}` and the levels decouple from the raw scores in useful ways — on this run, 25 of 31 assessments scored 0.7-0.9 but were labeled `level: low`, and the one confirmed drift at 0.834 was labeled `level: high` with `basis: "finding is backed by explicit doc/spec excerpts that also align semantically"`. That's a much richer signal than a scalar confidence would provide. **The RFC should formalize this layering as the output contract** — score + level + basis, not just score.

3. **Basis strings should be machine-enumerable.** The `basis` field is a free-text explanation today. Converting it to an enum (or a small tagged structure like `{matched_keywords: [...], semantic_similarity: 0.83, pragmatic_judgment_applied: false}`) would let downstream tooling route findings to human review based on why the model thinks what it thinks, not just how strongly.

4. **A deterministic historical-container post-filter is a concrete mitigation.** Detect CHANGELOG section headers (`## [\d+.\d+.\d+]` etc.), release notes, migration docs. Findings whose cited excerpt falls inside such a container get a confidence penalty (e.g., `level: medium → low`, or `score *= 0.5`). This is cheap, explainable, and addresses the class of failure that bigger models don't fix.

5. **Two-model adjudication is a viable cheap/accurate split.** Use the 9B as a fast bulk pass, re-run anything labeled `semantic_conflict` or `confirmed drift` against the 35B, and treat disagreements as human-review signals. On this run:
   - 9B conflicts re-adjudicated by 35B: 1 FP → 0 (35B overrides 9B correctly)
   - 9B confirmed drift re-adjudicated by 35B: 1 FP → 1 FP (no improvement, both wrong)

   So the two-model pattern works for `conflict` but not for historical-container drift. Worth documenting in the RFC which finding categories benefit from model re-adjudication vs which don't.

Happy to share the raw JSON for both runs — it's ~500KB total and useful as regression fodder for #289 (separate issue about check-terminology's false-positive rate).

**Comment by autholykos** [comment]
## Correction + NVIDIA Nemotron re-adjudication

The previous "35B re-adjudication" comment had three attribution errors I need to correct before adding a third data point (an NVIDIA Nemotron model). The substantive conclusions change as a result.

### Corrections to the prior 35B comment

**Correction 1 — `check-doc-drift` is embedding-only; the adjudicator model is not invoked.**

I re-ran with Qwen3.5-35B and with Nemotron using the exact same index. The drift JSON output files are **byte-identical across all three models** (same MD5 for Qwen3.5-9B, Qwen3.5-35B, and Nemotron). That is only possible if `check-doc-drift` never consults `runtime.analysis` at all — drift is apparently produced entirely from the `nomic-embed-text-v1.5` embeddings + pituitary's own structural code.

Separately verified by substituting a nonexistent model name into the override config and re-running: `check-compliance` fails fast with a `dependency_unavailable` 404 from the local model router, while `check-doc-drift` (if I had kept it in scope) would have completed normally because it never calls the analysis endpoint.

This means my prior sentences about "the 35B still produced 1 false positive on check-doc-drift" were **misattribution** — every model produces that same CHANGELOG finding because the model isn't consulted. A drift result cannot be used to compare adjudicator quality at all, in either direction.

**Correction 2 — The 9B's drift confidence was `0.834`, not `0.95`.**

My side-by-side table listed the 9B confirmed drift at `0.95` and the 35B at `0.834`. Both are actually `0.834` — they must be, since the JSON is byte-identical. The `0.95` in my prior comment belonged to the `check-compliance` `semantic_conflict` on the reference doc, not to drift. I conflated the two.

**Correction 3 — The layered `confidence: {level, score, basis}` structure is not a 35B-specific behaviour.**

I claimed the 35B emitted layered confidence while the 9B emitted scalar. That was wrong. The layered shape is populated by pituitary's own drift code path, not by the LLM's structured output — the 9B's drift_item also carries `{level: "high", score: 0.834, basis: "finding is backed by explicit doc/spec excerpts that also align semantically"}`. So my RFC point #2 ("the 35B's confidence layer is actually well-calibrated already — formalize this layering as the output contract") should be rephrased: pituitary *already* emits layered confidence at the drift layer across models. The RFC point I meant to make is about `check-compliance`, where the layered shape is not currently present and would be valuable.

### NVIDIA Nemotron re-adjudication — same corpus, same index

Re-ran with a Nemotron model (NVIDIA family) as `runtime.analysis.model`. Everything else identical (same index, same chunks, same embedder, same terminology policies, same 26-spec source set, AST inference on).

| | Qwen3.5-9B | Qwen3.5-35B | Nemotron |
|---|---|---|---|
| `check-compliance` compliant | 1 | 1 | 1 |
| `check-compliance` **conflicts** | **1 FP @ 0.95** | **0** ✅ | **0** ✅ |
| `check-compliance` unspecified | 25 | 25 | 25 |
| `check-compliance` MD5 of result | `6d7c0ac…` | `813e14f…` | `813e14f…` (identical to 35B) |
| `check-doc-drift` drift_items | 1 @ 0.834 | 1 @ 0.834 | 1 @ 0.834 |
| `check-doc-drift` MD5 of result | `daaabd6…` | `daaabd6…` | `daaabd6…` (identical to all) |
| Wall-clock (compliance, same diff) | — | 11.6s | 16.1s |

Both the Qwen3.5-35B and the Nemotron correctly rejected the 9B's `semantic_conflict` on the CLI-subcommand pair. The two non-9B adjudicators produce byte-identical responses because pituitary's `check-compliance` result envelope has no model-provenance field — when two adjudicators return the same structural answer, there is no field in the JSON that would distinguish them. I'd treat this as a small diagnostics gap: correct same-answer results should still be distinguishable at audit time. Happy to file separately if useful.

Note on speed: the Nemotron is roughly in the Qwen3.5-35B range on this workload (16s vs 12s for the same ~2-input compliance call). The Nemotron family routing on this MLX setup is comparable to Qwen3.5-35B here.

### What this means for the RFC

1. **Adjudicator choice matters for `check-compliance` but is inert for `check-doc-drift`.** Any RFC prose about "re-adjudicating drift findings" needs to either (a) become a request to route drift through `runtime.analysis` as well, or (b) explicitly scope adjudicator selection to `check-compliance`. Today only (b) is accurate.

2. **Two-model adjudication pattern holds, but only on the compliance path.** 9B as a fast bulk pass → Nemotron or 35B for anything labeled `semantic_conflict` works cleanly on this corpus: 1 9B FP → 0 on both larger adjudicators. The pattern does *not* help on the CHANGELOG historical-container drift finding because that finding is produced without ever calling an analysis model.

3. **Historical-container drift is a deterministic-filter problem, not a model-scale problem.** Because drift is embedding-only, the CHANGELOG release-narration false positive cannot be fixed by upgrading the analysis model. It needs either (a) pituitary-side structural awareness of changelog/release-notes containers, or (b) drift findings that fall inside those containers being downgraded by a deterministic post-filter, or (c) routing drift findings through `runtime.analysis` for a semantic second opinion before promoting them to `drift_items`.

4. **Adjudicator policy on this setup.** For future pituitary runs on this corpus, the operator-level policy is: Qwen3.5-9B as fast bulk-pass analyzer, **Nemotron as the adjudicator for `semantic_conflict` and `confirmed drift` findings** — not Qwen3.5-35B. This is an operator-policy choice (model family preference), not an empirical ranking on the one disagreement tested here, where the 35B and Nemotron tied at "correct".

5. **Two small upstream diagnostics gaps worth filing** (separate issues, not this RFC, unless useful here):
   - `check-compliance` result envelope has no `runtime.analysis.model` / `runtime.analysis.profile` field, so two adjudicators producing the same answer are indistinguishable at MD5 and at audit.
   - `check-doc-drift` does not call the analysis model at all. This should be documented clearly, or made configurable, so operators don't assume that swapping `runtime.analysis.model` will affect drift findings.

Raw JSON for all three runs is available on request.
